android/src/main/java/com/reactlibrary/RNCarnivalModule.java

```diff
@@ -40,524 +38,518 @@
-        public void onSuccess() {
-          promise.resolve(true);
+        return map;
+    }
+
+    private static WritableArray convertJsonToArray(JSONArray jsonArray) throws JSONException {
+        WritableArray array = new WritableNativeArray();
+
+        for (int i = 0; i < jsonArray.length(); i++) {
+            Object value = jsonArray.get(i);
+            if (value instanceof JSONObject) {
+                array.pushMap(convertJsonToMap((JSONObject) value));
+            } else if (value instanceof JSONArray) {
+                array.pushArray(convertJsonToArray((JSONArray) value));
+            } else if (value instanceof Boolean) {
+                array.pushBoolean((Boolean) value);
+            } else if (value instanceof Integer) {
+                array.pushInt((Integer) value);
+            } else if (value instanceof Double) {
+                array.pushDouble((Double) value);
+            } else if (value instanceof String) {
+                array.pushString((String) value);
+            } else {
+                array.pushString(value.toString());
+            }
         }
+        return array;
+    }
 
-        @Override
-        public void onFailure(Error error) {
-          promise.reject(ERROR_CODE_TRACKING, error.getMessage());
+    private static JSONObject convertMapToJson(ReadableMap readableMap) throws JSONException {
+        JSONObject object = new JSONObject();
+        ReadableMapKeySetIterator iterator = readableMap.keySetIterator();
+        while (iterator.hasNextKey()) {
+            String key = iterator.nextKey();
+            switch (readableMap.getType(key)) {
+                case Null:
+                    object.put(key, JSONObject.NULL);
+                    break;
+                case Boolean:
+                    object.put(key, readableMap.getBoolean(key));
+                    break;
+                case Number:
+                    object.put(key, readableMap.getDouble(key));
+                    break;
+                case String:
+                    object.put(key, readableMap.getString(key));
+                    break;
+                case Map:
+                    object.put(key, convertMapToJson(readableMap.getMap(key)));
+                    break;
+                case Array:
+                    object.put(key, convertArrayToJson(readableMap.getArray(key)));
+                    break;
+            }
         }
-      });
-    } catch (URISyntaxException e) {
-      promise.reject(ERROR_CODE_TRACKING, e.getMessage());
+        return object;
     }
-  }
-
-  @ReactMethod
-  public void trackPageview(String url, ReadableArray tags, final Promise promise) {
-    try {
-      List<String> convertedTags = null;
-      if(tags != null) {
-        convertedTags = new ArrayList<>();
-        for (int i = 0; i < tags.size(); i++) {
-          convertedTags.add(tags.getString(i));
+
+    private static JSONArray convertArrayToJson(ReadableArray readableArray) throws JSONException {
```

Comment:
I would love to move this out someday. Just to clean it up and maybe opensource it. It's quite a common task to convert from JSON to those `ReadableArray` arrays and vice-versa.

Reply:
Sounds good, could be a good change for a later release.